cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/MetricsMessageAggregator.java

```diff
@@ -0,0 +1,278 @@
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * An aggregator for metrics that aggregates metrics on 2 levels:
+ * <ol>
+ *   <li>Combine metrics across all {@link MetricValues} with the same tags, retaining the latest timestamp.</li>
+ *   <li>Inside each {@link MetricValues}, combine {@link MetricValue} with the same name.</li>
+ * </ol>
+ */
+class MetricsMessageAggregator {
+
+  private static final Logger LOG = LoggerFactory.getLogger(
+      MetricsMessageAggregator.class);
+
+  private final Map<String, String> cachemap;
+  private final ResettableByteArrayInputStream decodingInputStream;
+  private final StringCachingDecoder decoder;
+  private final DatumReader<MetricValues> reader;
+  private final Schema schema;
+  private final ByteArrayOutputStream encodeOutputStream;
+  private final BinaryEncoder encoder;
+  private final DatumWriter<MetricValues> recordWriter;
+  private final long aggregationWindowSeconds;
+
+  MetricsMessageAggregator(SchemaGenerator schemaGenerator,
+      DatumWriterFactory writerFactory, DatumReaderFactory readerFactory,
+      long aggregationWindowSeconds) {
+    TypeToken<MetricValues> metricValueType = TypeToken.of(MetricValues.class);
+    try {
+      this.schema = schemaGenerator.generate(metricValueType.getType());
+    } catch (UnsupportedTypeException e) {
+      throw new RuntimeException("Failed to generate schema for MetricValues",
+          e);
+    }
+    reader = readerFactory.create(metricValueType, schema);
+    recordWriter = writerFactory.create(metricValueType, schema);
+    this.cachemap = new HashMap<>();
+    this.decodingInputStream = new ResettableByteArrayInputStream();
+    this.decoder = new StringCachingDecoder(
+        new BinaryDecoder(decodingInputStream), cachemap);
+    this.encodeOutputStream = new ByteArrayOutputStream(1024);
+    this.encoder = new BinaryEncoder(encodeOutputStream);
+    this.aggregationWindowSeconds = Math.max(1, aggregationWindowSeconds);
+  }
+
+  public Iterator<Message> aggregate(Iterator<Message> input) {
+    cachemap.clear();
+    if (!input.hasNext()) {
+      return input;
+    }
+
+    Map<MetricsMapKey, MetricsAggregator> aggregated = new HashMap<>();
+    String[] messageId = new String[1];
+    long currentTimestamp = System.currentTimeMillis() / 1000;
+    // MetricValues will be aggregated across windows of size "aggregationWindowSeconds".
+    // We align the current time to the end of the last widow. This is done since in most cases,
+    // it is expected that all the MetricValues were emitted very recently. The offset ensures
+    // the last window completely lies in the past and maximizes the chance that all metrics lie within it.
+    long offset = (currentTimestamp % aggregationWindowSeconds) + 1;
+
+    input.forEachRemaining(message -> {
```

Comment:
This method is kind of too long. I see that it can at least broken down into aggregation and transform to write out.

Reply:
Broke the method into smaller pieces.